sphinx_needs/lsp/esbonio.py

```diff
@@ -0,0 +1,390 @@
+        if os.path.exists(doc_path + ".rst"):
+            doc_path = doc_path + ".rst"
+        elif os.path.exists(doc_path + ".rest"):
+            doc_path = doc_path + ".rest"
+        else:
+            return []
+        doc_uri = pathlib.Path(doc_path).as_uri()
+
+        # get the need definition position (line, col) from file
+        with open(doc_path) as file:
+            source_lines = file.readlines()
+        # get the line number
+        line_count = 0
+        line_no = None
+        pattern = f":id: {need_id}"
+        for line in source_lines:
+            if pattern in line:
+                line_no = line_count
+                break
+            line_count = line_count + 1
+        if not line_no:
+            return []
+
+        # get line of directive (e.g., .. req::)
+        line_directive = None
+        pattern = f".. {need_type}::"
+        for line_count in range(line_no - 1, -1, -1):
+            if pattern in source_lines[line_count]:
+                line_directive = line_count
+                break
+        if not line_directive:
+            return []
+
+        pos = Position(line=line_directive, character=0)
+        return [Location(uri=doc_uri, range=Range(start=pos, end=pos))]
+
+
+def col_to_word_index(col: int, words: List[str]) -> int:
+    """Return the index of a word in a list of words for a given line character column."""
+    length = 0
+    index = 0
+    for word in words:
+        length = length + len(word)
+        if col <= length + index:
+            return index
+        index = index + 1
+    return index - 1
+
+
+def get_lines(ls: NeedlsFeatures, params: Union[CompletionContext, DefinitionContext, HoverContext]) -> List[str]:
+    """Get all text lines in the current document."""
+    text_doc = params.doc
+    ls.logger.debug(f"text_doc: {text_doc}")
+    source = text_doc.source
+    lines: List[str] = source.splitlines()
+    return lines
+
+
+def get_word(ls: NeedlsFeatures, params: Union[CompletionContext, DefinitionContext, HoverContext]) -> str:
+    """Return the word in a line of text at a character position."""
+    line_no, col = params.position
+    lines = get_lines(ls, params)
+    if line_no >= len(lines):
+        return ""
+    line = lines[line_no]
+    words = line.split()
+    index = col_to_word_index(col, words)
+    word: str = words[index]
+    return word
+
+
+def get_lines_and_word(ls: NeedlsFeatures, params: CompletionContext) -> Tuple[List[str], str]:
+    return (get_lines(ls, params), get_word(ls, params))
+
+
+def get_need_type_and_id(
+    ls: NeedlsFeatures, params: Union[DefinitionContext, HoverContext]
+) -> Tuple[Optional[str], Optional[str]]:
+    """Return tuple (need_type, need_id) for a given document position."""
+    word = get_word(ls, params)
+    for need in ls.needs_store.needs.values():
+        if need["id"] in word:
+            return (need["type"], need["id"])
+    return (None, None)
```

Comment:
would this not be better returning a `Optional[Tuple[str, str]]`?

as in-

```suggestion
def get_need_type_and_id(
    ls: NeedlsFeatures, params: Union[DefinitionContext, HoverContext]
) -> Optional[Tuple[str, str]]:
    """Return tuple (need_type, need_id) for a given document position."""
    word = get_word(ls, params)
    for need in ls.needs_store.needs.values():
        if need["id"] in word:
            return (need["type"], need["id"])
    return None
```

obviously you'll need to cascade that change everywhere it's called, but i think it might make your null-checks a little tidier

Reply:
Don't agree. This function needs to return a Tuple (None, None), not just None. Otherwise, it will make the code in other place where use this function more complicated.

Reply:
If you say so!

I'm curious which bits will become more complex?

The reason I suggest using Optional[Tuple] is that the 'optional' monad never needs to be distributed over the tuple elements. That is, you *never* have [Some, None] or [None, Some], you only have [Some, Some] or [None, None]. That is, the type you've chosen has invalid variants